route_guide/route_guide_server.cc

```diff
@@ -55,133 +66,226 @@ float GetDistance(const Point& start, const Point& end) {
+                   long right = std::max(lo.longitude(), hi.longitude());
+                   long top = std::max(lo.latitude(), hi.latitude());
+                   long bottom = std::min(lo.latitude(), hi.latitude());
+                   for (const Feature& f : feature_list_) {
+                     if (f.location().longitude() >= left
+                         && f.location().longitude() <= right
+                         && f.location().latitude() >= bottom
+                         && f.location().latitude() <= top) {
+                       // The response to the client.
+                       call.Write(f);
+                     }
+                   }
+                   // When writes are done just finish the call.
+                   call.Finish(Status::OK);
+                 });
+           }))
+        | Loop();
+  }
 
-    status = server->Serve<RouteGuide, Stream<Point>, RouteSummary>(
-        "RecordRoute",
-        [this](auto&& call) {
-          int point_count = 0;
-          int feature_count = 0;
-          float distance = 0.0;
-          Point previous;
-
-          system_clock::time_point start_time = system_clock::now();
-
-          call->OnRead(
-              [this, point_count, feature_count, distance, previous, start_time](
-                  auto* call, auto&& point) mutable {
-            if (point) {
-              point_count++;
-              if (!GetFeatureName(*point, feature_list_).empty()) {
-                feature_count++;
-              }
-              if (point_count != 1) {
-                distance += GetDistance(previous, *point);
-              }
-              previous = *point;
-            } else {
-              system_clock::time_point end_time = system_clock::now();
-              RouteSummary summary;
-              summary.set_point_count(point_count);
-              summary.set_feature_count(feature_count);
-              summary.set_distance(static_cast<long>(distance));
-              auto secs = std::chrono::duration_cast<std::chrono::seconds>(
-                  end_time - start_time);
-              summary.set_elapsed_time(secs.count());
-              call->WriteAndFinish(summary, Status::OK);
-            }
-          });
-        });
-
-    if (!status.ok()) {
-      return status;
-    }
+  // ServeRecordRoute() function implements the client streaming RPC. The
+  // client sends 10 Point's messages to the server. Each message consists of
+  // the Point's data stucture. The server responds with a unique message which
+  // contains the number of points, number of feautures found in the db,
+  // the resulting distance calculated by using all Point's coordinates and
+  // the resulting time.
+  auto ServeRecordRoute(Server* server) {
+    return server->Accept<
+               RouteGuide,
+               Stream<Point>,
+               RouteSummary>("RecordRoute")
+        | Map(Then([this](auto&& context) {
```

Comment:
What's in `context` here? If we don't use it, can we omit it?

Reply:
Definitely not:) the `context` object is the `TypedServerContext` structure (`stout/grpc/server.h`), which has `unique_ptr<ServerContext>` context. this context we use as an obligatory argument in the `Server::Handler(TypedServerContext<Request, Response>&& context)` function. Implementation of the `Server::Handler(...)` function is also in the `stout/grpc/server.h` file.